serving/serving.go

```diff
@@ -99,132 +119,138 @@ func (serv *FeatureServer) TrainingTestSplit(stream pb.Feature_TrainingTestSplit
+		TrainingSetVariant: splitContext.req.Id.GetVersion(),
+		TestSize:           splitContext.req.TestSize,
+		Shuffle:            splitContext.req.Shuffle,
+		RandomState:        int(splitContext.req.RandomState),
+	}
 
+	cleanupFunc, err := serv.createTrainTestSplit(trainTestSplitDef)
+	defer cleanupFunc()
 	if err != nil {
-		logger.Errorw("Failed to get training set iterator", "Error", err)
+		return fferr.NewInternalError(err)
+	}
+	train, test, err := serv.getTrainTestSplitIterators(trainTestSplitDef)
+	if err != nil {
+		splitContext.logger.Errorw("Failed to get training set iterator", "Error", err)
 		return err
 	}
-	defer dropViews()
 
-	*trainIterator = train
-	*testIterator = test
+	*splitContext.trainIterator = train
+	*splitContext.testIterator = test
 
-	initResponse := &pb.TrainingTestSplitResponse{
+	initResponse := &pb.BatchTrainTestSplitResponse{
 		RequestType: pb.RequestType_INITIALIZE,
-		TrainingTestSplit: &pb.TrainingTestSplitResponse_Initialized{
-			Initialized: true,
-		},
+		Result:      &pb.BatchTrainTestSplitResponse_Initialized{Initialized: true},
 	}
 
-	if err := stream.Send(initResponse); err != nil {
-		logger.Errorw("Failed to send init response", "error", err)
+	if err := splitContext.stream.Send(initResponse); err != nil {
+		splitContext.logger.Errorw("Failed to send init response", "error", err)
 		return err
 	}
 
 	return nil
 }
 
-func (serv *FeatureServer) handleSplitDataRequest(
-	stream pb.Feature_TrainingTestSplitServer,
-	req *pb.TrainingTestSplitRequest,
-	trainIterator *provider.TrainingSetIterator,
-	testIterator *provider.TrainingSetIterator,
-	isTestFinished *bool,
-	isTrainFinished *bool,
-	logger *zap.SugaredLogger,
-) error {
-
+func (serv *FeatureServer) handleSplitDataRequest(splitContext *splitContext) error {
 	var thisIter provider.TrainingSetIterator
-	switch req.GetRequestType() {
+	switch splitContext.req.GetRequestType() {
 	case pb.RequestType_TRAINING:
-		thisIter = *trainIterator
+		thisIter = *splitContext.trainIterator
 	case pb.RequestType_TEST:
-		thisIter = *testIterator
+		thisIter = *splitContext.testIterator
 	default:
 		return fmt.Errorf("invalid request type")
 	}
 
-	if thisIter.Next() {
-		sRow, err := serializedRow(thisIter.Features(), thisIter.Label())
-		if err != nil {
-			return err
-		}
+	rows := 0
+	trainingDataRows := make([]*pb.TrainingDataRow, 0)
 
-		response := &pb.TrainingTestSplitResponse{
-			RequestType:       req.GetRequestType(),
-			TrainingTestSplit: &pb.TrainingTestSplitResponse_Row{Row: sRow},
+	for rows < int(splitContext.req.BatchSize) {
+		if thisIter.Next() {
+			sRow, err := serializedRow(thisIter.Features(), thisIter.Label())
+			if err != nil {
+				return err
+			}
+			trainingDataRows = append(trainingDataRows, sRow)
+			rows++
+		} else {
+			// if we reach the end of the iterator mid-batch, we'll send the processed rows so far and end the iteration
+			serv.handleFinishedIterator(trainingDataRows, splitContext)
 		}
```

Comment:
Do we need to `break` out of the loop in the `else` clause to avoid an infinite loop?